Configurable ambient, visit and removal sounds

ArseniyShestakov opened the pull request:
More details about H3 ambient sound implementation available on mantis:
https://bugs.vcmi.eu/view.php?id=2368
Other issues fixed:
https://bugs.vcmi.eu/view.php?id=814
https://bugs.vcmi.eu/view.php?id=1630

DjWarmonger commented on client/CPlayerInterface.cpp:
```diff
@@ -2813,3 +2814,83 @@ void CPlayerInterface::showWorldViewEx(const std::vector<ObjectPosInfo>& objectP
+
+		auto obj = tt->topVisitableObj(pos == tile);
+		if(obj && obj->ID != Obj::HERO)
+			updateObjects(obj->id);
```
id is unique identifier of objetc instance. Shouldn't sounds be differentiated by <ID, subID> pair?

ArseniyShestakov replied:
More likely I'll differentiate them by sound filename later on as I don't think we want multiple channels of same sound effect playing simultaneously even if multiple objects with same ambient present.

ArseniyShestakov commented:
So I'm finished checking map objects and made full list:
https://gist.github.com/ArseniyShestakov/616dd31231d5ac5942ad

alexvins commented on config/objects/dwellings.json:
```diff
@@ -2,6 +2,7 @@
 	"creatureGeneratorCommon" : {
 		"index" :17,
 		"handler": "dwelling",
+		"ambientSound" : "LOOPSWOR",
```
As you adding sounds on sybtype level, do not add it on top level. Add default sound in base struct below.

alexvins commented on config/objects/moddables.json:
```diff
@@ -183,19 +183,20 @@
+			"crystalCavern" : { "index" : 4, "rmg" : { "value" : 3500 }, "ambientSound" : "LOOPCRYS" },
+			"gemPond" :       { "index" : 5, "rmg" : { "value" : 3500 }, "ambientSound" : "LOOPGEMP" },
+			"goldMine" :      { "index" : 6, "rmg" : { "value" : 7000 }, "ambientSound" : "LOOPMINE" },
+			"abandoned" :	  { "index" : 7,}
```
redundant comma

alexvins commented on lib/IGameCallback.cpp:
```diff
@@ -65,7 +65,9 @@ void CPrivilagedInfoCallback::getTilesInRange( std::unordered_set<int3, ShashInt
 				int3 tilePos(xd,yd,pos.z);
 				double distance;
-				if(patrolDistance)
+				if(distanceFormula == 2) // ambient soud distance
```
distanceFormula shouldbe of some enum class type

alexvins replied:
Also extract calculation with formula selection to int3 method

alexvins commented on lib/IGameCallback.cpp:
```diff
@@ -47,11 +47,11 @@ void CPrivilagedInfoCallback::getFreeTiles (std::vector<int3> &tiles) const
 }
 
-void CPrivilagedInfoCallback::getTilesInRange( std::unordered_set<int3, ShashInt3> &tiles, int3 pos, int radious, boost::optional<PlayerColor> player/*=uninit*/, int mode/*=0*/, bool patrolDistance/*=false*/) const
+void CPrivilagedInfoCallback::getTilesInRange(std::unordered_set<int3, ShashInt3> &tiles, int3 pos, int radious, boost::optional<PlayerColor> player/*=uninit*/, int mode/*=0*/, int distanceFormula/*=0*/) const
```
I understand why we have unordered here (FoW calculation), but in most cases vector<int3> is enough and much faster. It would be good to have getTilesInRange with vector container too.

ArseniyShestakov commented:
Rebased on top of develop. Going to work on this now.

alexvins commented on road debug:
```diff
@@ -0,0 +1,291 @@
+OOOO  OOO  O OOOOOOOOOOOOOOOOO OOOOOOOOOOOOOOOO OOOOOOOOOOOOOOOOOO  OOOOOOOOO OO OO OOOOOOOOOO    OOOOOOOOOOOOOOOOOOOOOOOOOOOOOO  OOOOOOOOOOOOOO
```
this file should not be commied.

ArseniyShestakov replied:
Yeah accidentally added it while in rebase. Fixed.

alexvins commented on lib/mapObjects/CObjectHandler.h:
```diff
@@ -183,6 +186,7 @@ class DLL_LINKAGE CGObjectInstance : public IObjectInterface
 		}
 
 		h & pos & ID & subID & id & tempOwner & blockVisit & appearance;
+		h & ambientSound;
```
version check missing

alexvins commented on lib/mapObjects/CObjectHandler.cpp:
```diff
@@ -215,6 +216,8 @@ void CGObjectInstance::initObj(CRandomGenerator & rand)
 		break;
 	}
+
+	VLC->objtypeh->getHandlerFor(ID, subID)->setAmbient(this);
```
too much code duplication, if it too difficult to ensure CGObjectInstance::initObj call for all descendants move this initialization to CGameState::initMapObjects()

ArseniyShestakov replied:
Yeah I find out that doing that just after initObj call is way better since we initialize objects just in few places. Though there is problem that for whatever reason "CObjectClassesHandler::getHandlerFor" can't find handler for some objects and it's fail on runtime error.

ArseniyShestakov commented:
I had no internet for about 12 hours so implemented almost everything I wanted.
Also tested probably every map object and added all missing visit sounds.
Still need to rework playback and channel management code.

alexvins commented on lib/mapObjects/CObjectClassesHandler.cpp:
```diff
@@ -202,6 +203,12 @@ CObjectClassesHandler::ObjectContainter * CObjectClassesHandler::loadFromJson(co
+		obj->sounds.ambient = json["sounds"]["ambient"].String();
+	if(!json["sounds"]["visit"].String().empty())
+		obj->sounds.visit = json["sounds"]["visit"].String();
+
```
No need to check for empty string

alexvins commented on lib/mapObjects/MiscObjects.cpp:
```diff
@@ -713,7 +714,7 @@ void CGMine::initObj(CRandomGenerator & rand)
 	}
 
-	producedQuantity = defaultResProduction();
+	producedQuantity = defaultResProduction();;
```
typo

ArseniyShestakov commented:
So this one now rebased almost finished. Now sound IDs only stored in appropriate handlers and not in object instances since I guess this is what @DjWarmonger suggested two years ago. Also moved distance formula to ```int3```, implemented pickup sounds into schema, fixed few sounds that were still missing, etc.

Not perfect code, but works. Only problem is some glitches in sounds after battle is finished, but that might occur because some other code doesn't stop playback properly so I need to investigate.

DjWarmonger commented on lib/int3.h:
```diff
@@ -92,6 +92,31 @@ class int3
+		DIST_2DSQ
+	};
+
+	ui32 dist(const int3 & o, EDistanceFormula formula) const
```
Is this formula going to be used elsewhere? If in RMG or AI, I'd already implement it as a template + inlines to avoid additional overhead.
But probably it's not necessary to reuse it, anyway.

ArseniyShestakov replied:
I don't think it's will be heavily used since almost everything only need ```dist2d``` / ```dist2dSQ```. Manhattan distance used for patrol only and chebyshev only for ambient sounds that are not really performance critical, but it's still nice to explicitly specify what formula is used especially considering that we might need some other weird formulas in future.

alexvins commented on client/CPlayerInterface.cpp:
```diff
@@ -446,6 +448,15 @@ void CPlayerInterface::heroKilled(const CGHeroInstance* hero)
+	EVENT_HANDLER_CALLED_BY_CLIENT;
+	if(start && visitedObj->getVisitSound())
+	{
+		CCS->soundh->playSound(boost::str(boost::format("%s.wav") % visitedObj->getVisitSound().get()));
```
I do not like this. Make get[...]Sound return proper value, or pre-process inside `playSound`.

ArseniyShestakov replied:
I also want to investigate how this would work with mods. Actually I think our filesystem able to play identical music even if it's has different extension / container, but I will need to test it.

alexvins commented on client/CPlayerInterface.cpp:
```diff
@@ -2905,3 +2918,103 @@ void CPlayerInterface::showWorldViewEx(const std::vector<ObjectPosInfo>& objectP
+		}
+		// for nos steal a bit from CGameState::battleGetBattlefieldType which is super slow
+		// Also sound is for every special terrain on tile and not just one on top.
+		for(auto &obj : CGI->mh->map->objects)
```
Use CGI->mh->ttiles

alexvins commented on lib/mapObjects/CRewardableObject.h:
```diff
@@ -271,7 +269,11 @@ class DLL_LINKAGE CRewardableObject : public CArmedInstance
 		h & onEmpty;
 		h & visitMode;
-		h & soundID;
+		if(version < 778 && !h.saving)
```
Remove `&& !h.saving` condition. Initlialize soundID.

ArseniyShestakov replied:
Yeah I get ``` !h.saving``` likely make no sense, but we do have it in some other code. Should we get rid of it too?

alexvins replied:
Not in this branch

alexvins commented on lib/mapObjects/CObjectClassesHandler.h:
```diff
@@ -172,6 +190,10 @@ class DLL_LINKAGE AObjectTypeHandler : public boost::noncopyable
 			h & typeName;
 			h & subTypeName;
 		}
+		if(version >= 778)
```
The problem is that for any old save nearly all sounds will not work. You removed all hardcoded sounds, but do not provide defaults for old saves.

ArseniyShestakov replied:
That's true, but there no easy way to partially reload JSON...

henningkoehlernz commented on lib/int3.h:
```diff
@@ -109,6 +134,11 @@ class int3
+	//chebyshev distance used for ambient sounds (z coord is not used)
+	double chebdist2d(const int3 & o) const
+	{
+		return abs(std::max(abs(o.x) - abs(x), abs(o.y) - abs(y)));
```
This should be return std::max(abs(o.x - x), abs(o.y - y));

ArseniyShestakov commented:
So I improved the code, moved most to sound handler and also fixed not so obvious bug.
Now I just need to find out what's going wrong when hero surrender / flee.

ArseniyShestakov commented:
So just fixed problem with sound by proper locking.
Now everything almost ready except for compatibility.

ArseniyShestakov commented on client/CPlayerInterface.cpp:
```diff
@@ -2905,3 +2914,52 @@ void CPlayerInterface::showWorldViewEx(const std::vector<ObjectPosInfo>& objectP
+		{
+			auto obj = ttObj.obj;
+			if(!obj || !obj->getAmbientSound())
+				continue;
```
This one I especially dislike and this check doesn't help. Problem is that map handler information might sill have object that already removed.

I not sure how exactly to avoid that.

alexvins replied:
Copy sound-related data to `TerrainTileObject`.

ArseniyShestakov commented:
So now everything is ready except for the fact that loaded games won't have visit sounds.

DjWarmonger commented:
Okay, but lack of sounds for old saves is not a regression in any way. If that was merged few months back, nobody would even remember by now :P

dydzio0614 commented:
I think the deal are saves from bug reports, though I guess it does not affect anybody else than people who check bugs. I personally do not mind lacking visit sounds when checking bug reports.

alexvins commented:
@DjWarmonger please fix commit history

DjWarmonger commented:
That's exactly what I was trying to do this morning :v

DjWarmonger commented:
How can I do that?

alexvins commented:
1) DO NOT do `git pull`, it will merge remote changes to yours. (but you can do `git pull --ff-only`)
2) do `fetch`, then `rebase`

alexvins commented:
* now you need to forced rebase thus branch over develop. and (re)push

alexvins commented:
The only use of `git pull` in our workflow is 'git pull --ff-only' for develop branch is there **are no local changes** and **current local branch is also develop**

DjWarmonger commented:
But I always get this message:

```
git.exe push --progress "origin" feature/ambientSounds:feature/ambientSounds

To https://github.com/vcmi/vcmi/
! [rejected]          feature/ambientSounds -> feature/ambientSounds (non-fast-forward)
error: failed to push some refs to 'https://github.com/vcmi/vcmi/'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

alexvins commented:
should be `git.exe push --force-with-lease ...`. `force-with-lease` allows to rewrite commit history

alexvins commented:
In Tortoisegit it is `Force: may discard known changes`

DjWarmonger commented:
I think it works! Thanks.

alexvins commented:
* By default `push` do not overwrite commit history
* if you want do to that add f.e. `--force-with-lease`
* for protected branches (`develop` and `master` in out repo) history overwrite is forbidden entirely.
